theurgy.lic

```diff
@@ -1,186 +1,449 @@
-    fput "stow #{GameObj.left_hand.noun}" if GameObj.left_hand.name != 'Empty'
+  def inside_container?(item_name)
+    inside?(item_name, @theurgy_supply_container)
+  end
 
-    forage? 'dirt'
+  def put_in_container(item_name)
+    bput("put #{item_name} in my #{@theurgy_supply_container}",
+         'You put', 'What were you referring to')
+  end
 
-    get_from_container(@water_holder)
-    bput('commune eluned', 'completed this commune too recently', 'You grind some dirt in your fist')
-    put_in_container(@water_holder)
-    fput('drop dirt') if [left_hand, right_hand].include?('dirt')
+  def get_from_container(item_name)
+    bput("get #{item_name} from my #{@theurgy_supply_container}",
+         'You get', 'I could not find', 'What were you referring to')
+  end
+    
+  def have_holy_water?
+    inside?('holy water', @water_holder)    
   end
 
-  def comm_tamsine
-    if inside?('holy water', @water_holder)
-      pause 1
-      waitrt?
-      get_from_container(@water_holder)
-      fput "sprinkle #{@water_holder} on #{checkname}"
-      bput('commune tamsine', 'completed this commune too recently', 'You feel warmth spread throughout your body')
-      put_in_container(@water_holder)
-    else
-      Flags.delete('theurgy-tamsine')
-      nil
+  def buy_supplies
+    items_to_buy = (@rituals + @communes).map { |rc| rc[:items] }.flatten
+    items_to_buy.uniq! { |item| item[:name] }
+    items_to_buy.reject! { |item| item[:on_hand] }
+    items_to_buy.sort_by { |item| item[:shop]['id'] }
+    
+    ensure_copper_on_hand(600, @hometown) unless items_to_buy.empty?
+
+    items_to_buy.each do |item|
+      walk_to(item[:shop]['id'])
+      
+      if item[:purchase_method]
+        send(item[:purchase_method])
+      else
+        buy_item(item[:shop]['id'], item[:name])
+      end
+      put_in_container(item[:name])
     end
   end
 
-  def comm_kert
-    @hold_research = nil
-    if inside?('holy oil', @theurgy_supply_container)
-      get_from_container('holy oil')
-    else
-      case bput("get oil from #{@theurgy_supply_container}", 'holy oil', 'some oil', 'I could not find', 'What were you referring to')
-      when 'I could not find', 'What were you referring to'
-        return
-      when 'some oil'
-        research_check
-        fput 'prep bless'
-        pause 5
-        bput('cast oil', 'leaving it somehow changed', 'nothing else seems to happen', 'backfires')
+  def buy_taffelberries
+    bput('order 14',
+         'puts some sugar-dipped taffelberries on the bar',
+         'You don\'t have enough money to afford that')
+    fput 'get Taffelberries'
+  end
+
+  def buy_parchment
+    fput 'read placard'
+    pause 1
+    fput 'order 4 from monk'
+    fput 'unroll my parchment'
+  end
+  
+  def perform_next_action
+    perform_next_ritual unless perform_next_commune
+  end
+
+  ### COMMUNES ###
+  def perform_next_commune
+    return false if @communes.empty? || !can_commune?
```

Comment:
Should the `false if @communes.empty?` clause be part of `can_commune?`?

Reply:
You should probably remove the double negations as well, `return false unless can_commune?`

Reply:
Done.